src/extension/declarative/signature.rs

```diff
@@ -0,0 +1,257 @@
+                })
+            }
+        };
+
+        let ty = self.type_decl.make_type(ext, scope, registry, op_params)?;
+        let ty = Type::new_extension(ty);
+
+        Ok(itertools::repeat_n(ty, n))
+    }
+}
+
+impl Serialize for SignaturePortDeclaration {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let len = if is_default(&self.repeat) { 2 } else { 3 };
+        let mut seq = serializer.serialize_seq(Some(len))?;
+        seq.serialize_element(&self.description)?;
+        seq.serialize_element(&self.type_decl)?;
+        if !is_default(&self.repeat) {
+            seq.serialize_element(&self.repeat)?;
+        }
+        seq.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for SignaturePortDeclaration {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        struct TypeParamVisitor;
+        const EXPECTED_MSG: &str = "a 2-element list containing a type parameter name and id";
+
+        impl<'de> serde::de::Visitor<'de> for TypeParamVisitor {
+            type Value = SignaturePortDeclaration;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str(EXPECTED_MSG)
+            }
+
+            fn visit_seq<A: serde::de::SeqAccess<'de>>(
+                self,
+                mut seq: A,
+            ) -> Result<Self::Value, A::Error> {
+                let description = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(0, &EXPECTED_MSG))?;
+                let type_decl = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(1, &EXPECTED_MSG))?;
+                let repeat = seq.next_element()?.unwrap_or_default();
+                Ok(SignaturePortDeclaration {
+                    description,
+                    type_decl,
+                    repeat,
+                })
+            }
+        }
+
+        deserializer.deserialize_seq(TypeParamVisitor)
+    }
+}
+
+/// A number of repetitions for a signature's port definition.
+///
+/// This value may be either:
+/// - A number, indicating a repetition of the port that amount of times.
+/// - A parameter identifier, to use as the repetition number.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+enum PortRepetitionDeclaration {
+    /// A constant number of repetitions for the port definition.
+    Count(usize),
+    /// An (integer) operation parameter identifier to use as the number of repetitions.
+    Parameter(SmolStr),
+}
+
+impl Default for PortRepetitionDeclaration {
+    fn default() -> Self {
+        PortRepetitionDeclaration::Count(1)
+    }
+}
+
+/// A type declaration used in signatures.
+///
+/// TODO: The spec definition is more complex than just a type identifier,
+/// we should be able to support expressions like:
```

Comment:
yes - this would form a part of the type scheme definition language (the
fragment for types)